### PR TITLE
Auto checkpoint #106

### PR DESCRIPTION
**Pull request by yadudoc**
This brings in automatic checkpointing as described in #106 
Added following automated checkpointing methods

* task_exit
* dfk_exit
* periodic (with user specified intervals)

This also contains user docs, and tests fort the functionality added.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -17,5 +17,6 @@ os:
 script:
     - export PARSL_TESTING="true"
     - pip install -r test-requirements.txt
-    - nosetests -v parsl/tests/test_threads parsl/tests/test_data parsl/tests/test_checkpointing
+    - nosetests -v parsl/tests/test_threads parsl/tests/test_data
+    - for test in parsl/tests/test_checkpointing/test*; do nosetests -v $test; done
```
Comment (annawoodard) [inline .travis.yml]: Why is this necessary? Nose will traverse through directory levels. Also, maybe we should change `-v` -> `-vx` so things fail faster.

File: docs/devguide/dev_docs.rst [hunk 2]
```diff
@@ -96,7 +96,7 @@ DataFlowKernel
 ==============
 
 .. autoclass:: parsl.dataflow.dflow.DataFlowKernel
-   :members:  __init__, submit, cleanup, load_checkpoints, config, handle_update, launch_task, _count_deps, _count_all_deps, sanitize_and_wrap, _load_checkpoints
+   :members:  __init__, submit, cleanup, load_checkpoints, checkpoint, config, handle_update, launch_task, _count_deps, _count_all_deps, sanitize_and_wrap, _load_checkpoints
```
Comment (annawoodard) [inline docs/devguide/dev_docs.rst]: Isn't this just all the members? Unless we're trying to purposely exclude members, I think we should consider
```
:members:
```
instead of
```
:members:  __init__, submit, cleanup, load_checkpoints, checkpoint, config, handle_update, launch_task, _count_deps, _count_all_deps, sanitize_and_wrap, _load_checkpoints
```
because it is easier to maintain so that we don't have to remember to add a member here if we add one to the code. That applies everywhere...

File: docs/userguide/checkpoints.rst [hunk 3]
```diff
@@ -18,46 +18,92 @@ will checkpoint all results that have updated since the last checkpoint. When lo
...
-``DataFlowKernel.checkpoint()`` function is invoked. The checkpoint is written
-out to the ``runinfo/RUN_ID/checkpoint/`` directory.
+Checkpointing works in the following modes :
 
-Example code follows:
+1. ``task_exit`` : In this mode, each app is checkpointed when it has completed or
+   failed (after retries if enabled). This mode reduces the risk of losing information
+   from completed tasks to a minimum.
+
+   >>> config["globals"]["checkpoint"] = 'task_exit'
```
Comment (annawoodard) [inline docs/userguide/checkpoints.rst]: Let's consider reserving the use of `>>>`. It is the Python interpreter prompt and the standard way of running [doc tests](https://docs.python.org/2/library/doctest.html) so people may think they should be able to cut and paste directly into a prompt to test something. By not using this prompt here we avoid possible confusion and could eventually support doc tests.

File: parsl/dataflow/dflow.py [hunk 4]
```diff
@@ -94,12 +96,27 @@ def __init__(self, config=None, executors=None, lazyErrors=True, appCache=True,
             self.lazy_fail = self._config["globals"].get("lazyErrors", lazyErrors)
             self.fail_retries = self._config["globals"].get("retries", retries)
             self.flowcontrol = FlowControl(self, self._config)
+            self.checkpoint_method = self._config["globals"].get("checkpoint",
+                                                                 checkpointMethod)
+            if self.checkpoint_method == "periodic":
+                print("Setting periodic")
```
Comment (annawoodard) [inline parsl/dataflow/dflow.py]: I vote not to mix print and logging statements

File: parsl/dataflow/dflow.py [hunk 5]
```diff
@@ -46,7 +45,7 @@ class DataFlowKernel(object):
     """
 
     def __init__(self, config=None, executors=None, lazyErrors=True, appCache=True,
-                 rundir=None, retries=0, checkpointFiles=None):
+                 rundir=None, retries=0, checkpointFiles=None, checkpointMethod=None):
```
Comment (annawoodard) [inline parsl/dataflow/dflow.py]: can we call this `checkpointMode` instead of `checkpointMethod`? This gives 'method' a double meaning so it is confusing.

File: parsl/dataflow/dflow.py [hunk 6]
```diff
@@ -587,15 +621,18 @@ def checkpoint(self):
...
+                    logger.debug("Task {} checkpointed".format(task_id))
+
+        self.checkpointed_tasks += count
 
-        end = time.time()
         if count == 0:
-            logger.warn(
-                'No tasks checkpointed, please ensure caching is enabled')
+            if self.checkpointed_tasks == 0:
+                logger.warn("No tasks checkpointed, please ensure caching is enabled")
```
Comment (annawoodard) [inline parsl/dataflow/dflow.py]: Can we just check if caching is enabled automatically and return an error if someone tries to instantiate the DFK this way?

File: parsl/tests/test_checkpointing/test_dfk_exit.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
+''' Testing python apps
```
Comment (annawoodard) [inline parsl/tests/test_checkpointing/test_dfk_exit.py]: This is not a helpful docstring 😄, I prefer to just omit it in this case.

File: parsl/dataflow/flow_control.py [hunk 8]
```diff
@@ -141,8 +151,94 @@ def close(self):
...
+
+    '''
+
+    def __init__(self, callback, *args, interval=5):
+        ''' Initialize the flowcontrol object
+        We start the timer thread here
+
+        Args:
+             - dfk (DFK object) : DFK object to track parsl progress
+             - config (Dict) : Config dict structure
```
Comment (annawoodard) [inline parsl/dataflow/flow_control.py]: `DFK object` -> `DataFlowKernel`
`Dict` -> `dict`

If you use the actual class names sphinx will cross-reference them (it may need some tweaking to the config, I will submit it as a PR).

File: parsl/dataflow/flow_control.py [hunk 9]
```diff
@@ -141,8 +151,94 @@ def close(self):
         self._thread.join()
 
 
+class Timer(object):
+    '''This timer is a simplified version of the FlowControl timer.
+    This timer does not employ notify events.
+
+    This is based on the following logic :
+
+    >>> BEGIN (INTERVAL, THRESHOLD, callback) :
```
Comment (annawoodard) [inline parsl/dataflow/flow_control.py]: Same comment about the Python interpreter prompt.

Also, this is just personal preference, so no pressure to actually change this: I prefer to avoid toy code in documentation in most cases. The actual code this toy code is describing is only ~2x as long, and Python syntax is so simple it is basically toy code. But sometimes toy code cannot convey all of the proper details, so in the end you read the toy code and the real code and have to do the additional work of mapping between the two. So I prefer to use docs for exposition. Just my two cents.

File: .travis.yml (hunk 1 above)
Comment (yadudoc) [reply .travis.yml]: Agreed on -v -> -vx.
Nosetests is doing something unusual when running these tests from a directory. They pass when called one by one. This was easier that debugging what nose is doing.

File: docs/userguide/checkpoints.rst (hunk 3 above)
Comment (yadudoc) [reply docs/userguide/checkpoints.rst]: This is paste'able into the interpreter and is valid code. I don't understand the problem.
What I want to do is hilight code snippets assuming that users will copy paste it.

File: parsl/dataflow/dflow.py (hunk 6 above)
Comment (yadudoc) [reply parsl/dataflow/dflow.py]: This is harder. We can test for appCaching being enabled at the top_level, but we won't be able to find out if specific apps have caching enabled. This is why that warning is added.

Comment (annawoodard) [reply parsl/dataflow/dflow.py]: Sounds good, thanks for explaining.

File: parsl/dataflow/dflow.py [hunk 10]
```diff
@@ -94,12 +96,26 @@ def __init__(self, config=None, executors=None, lazyErrors=True, appCache=True,
             self.lazy_fail = self._config["globals"].get("lazyErrors", lazyErrors)
             self.fail_retries = self._config["globals"].get("retries", retries)
             self.flowcontrol = FlowControl(self, self._config)
+            self.checkpoint_method = self._config["globals"].get("checkpointMode",
```
Comment (annawoodard) [inline parsl/dataflow/dflow.py]: @yadudoc where we have properties can they always mirror the option? That way we don't have to keep track of the mapping?

Comment (annawoodard) [reply parsl/dataflow/dflow.py]: In other words `checkpoint_mode` instead of `checkpoint_method`
